lib/CL/clCommandFillImageKHR.c

```diff
@@ -0,0 +1,102 @@
+
+   The above copyright notice and this permission notice shall be included in
+   all copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+   IN THE SOFTWARE.
+*/
+
+#include <CL/cl_ext.h>
+
+#include "pocl_cl.h"
+#include "pocl_image_util.h"
+#include "pocl_shared.h"
+#include "pocl_util.h"
+
+CL_API_ENTRY cl_int
+POname (clCommandFillImageKHR) (
+    cl_command_buffer_khr command_buffer, cl_command_queue command_queue,
+    cl_mem image, const void *fill_color, const size_t *origin,
+    const size_t *region, cl_uint num_sync_points_in_wait_list,
+    const cl_sync_point_khr *sync_point_wait_list,
+    cl_sync_point_khr *sync_point,
+    cl_mutable_command_khr *mutable_handle) CL_API_SUFFIX__VERSION_1_2
+{
+  cl_int errcode;
+  _cl_recorded_command *cmd = NULL;
+
+  CMDBUF_VALIDATE_COMMON_HANDLES;
+
+  errcode = pocl_validate_fill_image (command_queue, image, fill_color, origin,
+                                      region);
+  if (errcode != CL_SUCCESS)
+    return errcode;
+
+  cl_uint4 fill_color_vec = *(const cl_uint4 *)fill_color;
+
+  size_t px = image->image_elem_size * image->image_channels;
+  char fill_pattern[16];
+  pocl_write_pixel_zero (fill_pattern, fill_color_vec,
+                         image->image_channel_order, image->image_elem_size,
+                         image->image_channel_data_type);
+
+  /* The fill color is:
+   *
+   * a four component RGBA floating-point color value if the image channel
+   * data type is NOT an unnormalized signed and unsigned integer type,
+   *
+   * a four component signed integer value if the image channel data type
+   * is an unnormalized signed integer type and
+   *
+   * a four component unsigned integer value if the image channel data type
+   * is an unormalized unsigned integer type.
+   *
+   * The fill color will be converted to the appropriate
+   * image channel format and order associated with image.
+   */
+
+  if (IS_IMAGE1D_BUFFER (image))
+    {
+      return POname (clCommandFillBufferKHR) (
+          command_buffer, command_queue, image->buffer, fill_pattern, px,
+          origin[0] * px, region[0] * px, num_sync_points_in_wait_list,
+          sync_point_wait_list, sync_point, mutable_handle);
+    }
+
+  char rdonly = 0;
+  errcode = pocl_create_recorded_command (
+      &cmd, command_buffer, command_queue, CL_COMMAND_FILL_IMAGE,
+      num_sync_points_in_wait_list, sync_point_wait_list, 1, &image, &rdonly);
+  if (errcode != CL_SUCCESS)
+    goto ERROR;
+
+  POCL_FILL_COMMAND_FILL_IMAGE;
```

Comment:
Reusing the buffered function as-is won't work very well - for example the spec specifically dictates that if buffer is not a valid command buffer the functions should return an error.

I'll take a stab at creating a shared implementation that is called with a flag to choose between the buffered and unbuffered variants. Should be a lot more manageable now that almost all relevant functions are down to roughly one page of code before macro expansions and once it's done I can get rid of the macros again,

I'm hesitant to reuse the full command struct creation code (and command node struct) since the needs of buffered and unbuffered commands are a bit different (queue index, sync points, list of related mem objects vs absolute device, events and implicitly enqueued commands for mem object migrations with the list of objects not actually retained in the command node). And I like having at least a chance of the compiler yelling at me if I accidentally try hand the wrong type of command to a function.